### ISSUE TITLE
Paket-ed UWP application fails to build with "Cannot resolve Assembly or Windows Metadata file 'System.Private.ServiceModel.dll'"

### ISSUE THREAD

**Issue by dbruning** [fsprojects/Paket#2531]
### Description

Trying to use Paket with a UWP application. I had previously logged #2465,  and the resolution of that allowed the convert-from-nuget command to complete.

However, after the command has completed, when I then try to then build the application, I get this error:
C:\Users\me\Documents\paket-uwp-repro\paket-uwp-repro\paket-uwp-repro.csproj : XamlCompiler error WMC1006: Cannot resolve Assembly or Windows Metadata file 'System.Private.ServiceModel.dll'
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Microsoft\WindowsXaml\v15.0\8.2\Microsoft.Windows.UI.Xaml.Common.targets(266,5): Xaml Internal Error error WMC9999: Type universe cannot resolve assembly: System.Private.ServiceModel, Version=4.1.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a.

(Sorry there's such a time delay between these 2 issues - I'm switching between projects)

### Repro steps

Create a new blank UWP application in Visual Studio
Add paket
Run paket convert-from-nuget
Try to build application

This can be reproduced with the same [Git repo](https://github.com/dbruning/paket-uwp-repro.git) as #2465 

### Expected behavior

Application has all required dependencies referenced, and builds successfully

### Actual behavior

Build error, as above

### Known workarounds

None

**Comment by dbruning** [comment]
I know bumps are unhelpful but I'd like some guidance on whether Paket officially supports UWP. I've refactored parts of my WPF codebase to support dotNet Core with the intention of eventually moving to UWP. Paket has been awesome in helping me sort out dependency issues but if I can't use it with UWP it kind of defeats the purpose for me.

**Comment by forki** [comment]
can you be a bit more specific on what is wrong in the repo? which references are wrong / missing?

**Comment by forki** [comment]
ok found something

**Comment by forki** [comment]
ok it seems this kind of project is not compatible. You are using MSBuild 15 without SDK but with

    <PropertyGroup>
      <RestoreProjectStyle>PackageReference</RestoreProjectStyle>
    </PropertyGroup>
    ...
    <ItemGroup>
      <PackageReference Include="Microsoft.NETCore.UniversalWindowsPlatform">
        <Version>5.4.0</Version>
      </PackageReference>
    </ItemGroup>

Unfortunatly if we convert this properly to modern paket and try to set the PackageReference in RAM then it doesn't work. I assume this is because there is plugin magic involved and this kind of thing doesn't work fully via MSBuild properties (cc @davkean)

I recommend you transform your initial UWP project to something that uses the new sdk and then run `paket convert-from-nuget` - it's a bit unfortunate that File/New Project creates such a kind of project in VS2017

**Comment by davkean** [comment]
UWP is using legacy project system, it doesn't read `PackageReference` using evaluation or targets, this work is being tracked by: https://github.com/dotnet/project-system/issues/2129.

**Comment by 949f45ac** [comment]
Hi there, I just stumbled upon this ticket yesterday as I was also trying to make a UWP project work with Paket. I've more or less just started working with VS projects, hence I'd be glad if you could clarify these points for me:
- Is there a chance to make this work in Paket as is, or do we have to wait for the dotnet/project-system ticket to be resolved? If it can be done in Paket, I'd be interested in trying my hand at contributing the missing parts, if you can roughly explain to me what they are.
- What do you mean by "uses the new sdk"?

Cheers

**Comment by forki** [comment]
It probably is possible to make it work in a very crippled way if we emit
the project references directly into the csproj.

But this will probably lead to new unknown issues and will be obsolete very
soon.

Am 02.08.2017 09:47 schrieb "dur" <notifications@github.com>:

**Comment by dbruning** [comment]
Thanks for your comments but I'm still unclear. Is there any version of UWP that can be made to work with Paket? If you could create some guidelines on getting Paket working at all with UWP then I would be happy to recreate my project following those guidelines & add my code back in.

I freely admit that I don't currently understand how the UWP csproj works (like @949f45ac I don't understand what "uses the new sdk" means), I'm happy to dive into learning that but still looking for a steer on whether Paket can currently be used at all on any UWP application. 

I just tried creating a new UWP app targetting 15063 and then doing paket convert-from-nuget. I ended up with a only Microsoft.NETCore.UniversalWindowsPlatform in packet.references and paket.dependencies, and then a huge paket.lock file and ~30 references that could not be found.

**Comment by 949f45ac** [comment]
I fiddled with this some today and my findings are roughly thus:
- A basic problem is that Paket works with TargetFramework / TargetFrameworkIdentifier / TargetFrameworkVersion to select dlls. However, UWP projects have none of these in their `csproj` normally, or at least my UWP project did not. This means Paket treats your project as "net40" and a lot of deps are just missed out on, because they do not target this.
- Now you can manually add TargetFramework=uap100 to your `csproj`. However, Paket does not know that "uap100" can use dlls built for a bunch of other frameworks, most notably "netcore50". (See [here](https://docs.microsoft.com/en-us/nuget/schema/target-frameworks) under "Precedence".) This means that we still do not get the correct assemblies.
(I looked at the normal build and it also pulls in "netstandard1.3" deps, even though that's not in the doc I linked.)
- Setting the "uap100" target framework also screws up the build process I think (app is taken for another kind of project), so the only route left is to edit the `csproj` by hand to directly reference the correct assemblies downloaded by Paket.
- I more or less tested this and it still did not work, failing at XAML compilation without a way for me to diagnose further, because the `MSBuild.failure.txt` file supposedly created in tmp was not there. It said "MSB4166: Child node 2 exited prematurely"

Overall, the feeling I got was that for some reason the whole UWP build is arcane. I did not look into it further (I'm not even sure it's all open source?), but it greatly differs from the other project I migrated to Paket. I mean, ordinarily it does not even list the referenced assemblies in Visual Studio (at least not for me) – they are only in `project.json`, and presumably the build script looks at that.

For ordinary kinds of project, where assemblies are referenced in `csproj`, there is a simple way to relatively sure success: You just have to end up with references to the same assemblies.

But in this case, it is ultimately unsurprising that it’s no good to add things like `mscorlib` and `System.dll` (which it will too demand of you eventually, once you have fixed `System.Private.ServiceModel` and all the others) to the UWP.csproj, because this is not even the normal way for them to be pulled into this.

**Comment by dbruning** [comment]
@949f45ac thanks for that analysis. I just spent the best part of a day reproducing what you had done and couldn't get any further or see any other workarounds.
Seems that Paket is just fundamentally incompatible with UWP. Wish I knew that before I spent the time migrating to Paket.
Looks like my options at this stage are to go back to Nuget, or to wait and hope that something changes with UWP such that Paket can be made to work with it. The issue referenced above - [dotnet/project-system#2129](https://github.com/dotnet/project-system/issues/2129) has a milestone of 15.5 but that milestone doesn't have a target date and is currently 2% complete.
I know there's a newer version of UWP coming later this year, but I can't find much information on it. Anyone know whether they're planning to pick up the new project system perchance?

**Comment by forki** [comment]
Sorry my friend. We didn't know the either. It looked very much like the
rest of the netstandard / netcore / new MSBuild world.

Am 23.08.2017 6:50 vorm. schrieb "Darren Bruning" <notifications@github.com

**Comment by dbruning** [comment]
Thanks @forki . I was thinking some more about this overnight - do you think support for UWP is something that's possible to add? Surely we could make paket do whatever Nuget does and just put the references into the .csproj directly. You mentioned this in a past comment:

> It probably is possible to make it work in a very crippled way if we emit the project references directly into the csproj.
>But this will probably lead to new unknown issues and will be obsolete very soon.

By "obsolete very soon" do you mean when UWP updates their project system? Because even then there will be people with legacy UWP projects, so the code would still be useful. 

Can you outline any other issues you can think of?

If I broadened my mind and learned F# and implemented this would it be a welcome PR?

Thanks, Darren.

**Comment by forki** [comment]
In theory we could emit PackageReferences tags directly into your csproj.
That would probably work fine.

But that would need someone to do it. And to fix all the bugs that will
come. Things like that never work first time.

And then when UWP updates, then everything is obsolete

Am 23.08.2017 10:02 nachm. schrieb "Darren Bruning" <
notifications@github.com>:

**Comment by dbruning** [comment]
Yeah that's all true. But strategically, if Paket wants to be better than Nuget, it has to at least support all the project types that Nuget does?
Given all that, would a PR be welcome?

**Comment by forki** [comment]
No we don't want to be better than nuget. That's not how I see the
mission.  We want to to solve our own issues and if that's good for others
than we are happy. Support for all project types is not an explicit goal.

Tbh I'm really not sure what the best way is here. I mean a pull request
would probably be merged, but it's definitely a lot of work. And the
question is if that's really worth it given that they will probably change
things soon.

Am 23.08.2017 10:17 nachm. schrieb "Darren Bruning" <
notifications@github.com>:

**Comment by dbruning** [comment]
OK, thanks for the guidance. I'll try to delay my UWP project until UWP vNext ships, and hope that they either pickup the new project system, or have suitable fixes in the old one so that implementing support in Paket isn't such a mission.

**Comment by 0x53A** [comment]
As I understand it, this affects Xamarin for both UWP (which no one cares about), and Android, which some people maybe care about.

**Comment by davkean** [comment]
@0x53A Above is https://github.com/dotnet/project-system/issues/2129.
